// phantomjs-adapter/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Element = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _class, _temp;

var _child_process = require('child_process');

var _phantomjsPrebuilt = require('phantomjs-prebuilt');

var _cssToXpath = require('css-to-xpath');

var _cssToXpath2 = _interopRequireDefault(_cssToXpath);

var _client = require('phantomjs-adapter/lib/client');

var _client2 = _interopRequireDefault(_client);

var _findScript = require('phantomjs-adapter/lib/find-script');

var _findScript2 = _interopRequireDefault(_findScript);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Element = exports.Element = function () {
  function Element(browser, data) {
    (0, _classCallCheck3.default)(this, Element);

    this.browser = browser;
    (0, _assign2.default)(this, data);
  }

  (0, _createClass3.default)(Element, [{
    key: 'click',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
        var _boundingClientRect, top, left, width, height;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _boundingClientRect = this.boundingClientRect;
                top = _boundingClientRect.top;
                left = _boundingClientRect.left;
                width = _boundingClientRect.width;
                height = _boundingClientRect.height;
                _context.next = 7;
                return this.browser.sendEvent('click', left + width / 2, top + height / 2);

              case 7:
                return _context.abrupt('return', _context.sent);

              case 8:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function click() {
        return _ref.apply(this, arguments);
      }

      return click;
    }()
  }, {
    key: 'fillIn',
    value: function () {
      var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(text) {
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.click();

              case 2:
                _context2.next = 4;
                return this.browser.sendEvent('keypress', text);

              case 4:
                return _context2.abrupt('return', _context2.sent);

              case 5:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function fillIn(_x) {
        return _ref2.apply(this, arguments);
      }

      return fillIn;
    }()
  }]);
  return Element;
}();

var _default = (_temp = _class = function () {
  function _default() {
    (0, _classCallCheck3.default)(this, _default);

    this.process = (0, _child_process.spawn)(_phantomjsPrebuilt.path, [this.constructor.scriptPath], {
      stdio: ['pipe', 'pipe', 'inherit']
    });
    this.client = new _client2.default(this.process);
  }

  (0, _createClass3.default)(_default, [{
    key: 'exit',
    value: function () {
      var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3() {
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.client.send('exit');

              case 2:
                return _context3.abrupt('return', _context3.sent);

              case 3:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function exit() {
        return _ref3.apply(this, arguments);
      }

      return exit;
    }()
  }, {
    key: 'open',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(url) {
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this.client.send('open', [url]);

              case 2:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function open(_x2) {
        return _ref4.apply(this, arguments);
      }

      return open;
    }()
  }, {
    key: 'evaluate',
    value: function () {
      var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(functionBody) {
        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this.client.send('evaluate', [functionBody]);

              case 2:
                return _context5.abrupt('return', _context5.sent);

              case 3:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function evaluate(_x3) {
        return _ref5.apply(this, arguments);
      }

      return evaluate;
    }()
  }, {
    key: 'sendEvent',
    value: function () {
      var _ref6 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6() {
        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        return _regenerator2.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this.client.send('sendEvent', [].concat(args));

              case 2:
                return _context6.abrupt('return', _context6.sent);

              case 3:
              case 'end':
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function sendEvent(_x4) {
        return _ref6.apply(this, arguments);
      }

      return sendEvent;
    }()
  }, {
    key: 'title',
    value: function () {
      var _ref7 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7() {
        return _regenerator2.default.wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this.evaluate('\n      return document.title;\n    ');

              case 2:
                return _context7.abrupt('return', _context7.sent);

              case 3:
              case 'end':
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function title() {
        return _ref7.apply(this, arguments);
      }

      return title;
    }()
  }, {
    key: 'find',
    value: function () {
      var _ref8 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee9(selector) {
        var sleep = function () {
          var _ref10 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee8(ms) {
            return _regenerator2.default.wrap(function _callee8$(_context8) {
              while (1) {
                switch (_context8.prev = _context8.next) {
                  case 0:
                    _context8.next = 2;
                    return new _promise2.default(function (resolve) {
                      setTimeout(resolve, ms);
                    });

                  case 2:
                  case 'end':
                    return _context8.stop();
                }
              }
            }, _callee8, this);
          }));

          return function sleep(_x8) {
            return _ref10.apply(this, arguments);
          };
        }();

        var _ref9 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

        var text = _ref9.text;
        var _ref9$wait = _ref9.wait;
        var wait = _ref9$wait === undefined ? 0 : _ref9$wait;
        var xpath, script, data, remainingWait;
        return _regenerator2.default.wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                xpath = typeof text === 'string' ? _cssToXpath2.default.parse(selector).where(_cssToXpath2.default.xPathBuilder.text().contains(text)).toXPath() : (0, _cssToXpath2.default)(selector);
                script = (0, _findScript2.default)(xpath);
                _context9.next = 4;
                return this.evaluate(script);

              case 4:
                data = _context9.sent;
                remainingWait = wait;

              case 6:
                if (!(!data && remainingWait > 0)) {
                  _context9.next = 15;
                  break;
                }

                _context9.next = 9;
                return sleep(100);

              case 9:
                remainingWait -= 100;
                _context9.next = 12;
                return this.evaluate(script);

              case 12:
                data = _context9.sent;
                _context9.next = 6;
                break;

              case 15:
                return _context9.abrupt('return', data ? new this.constructor.Element(this, data) : null);

              case 16:
              case 'end':
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function find(_x5, _x6) {
        return _ref8.apply(this, arguments);
      }

      return find;
    }()
  }, {
    key: 'logs',
    get: function get() {
      return this.client.logs;
    }
  }]);
  return _default;
}(), _class.Element = Element, _class.scriptPath = require.resolve('phantomjs-adapter/phantom'), _temp);

exports.default = _default;
